File: 06-FEA-3D-Elasticity/Matlab-Code/EvaluateFieldGradient.m
function [StressAtIntegrationPoints, StrainAtIntegrationPoints, ...
        StressAtCentroid, StrainAtCentroid,StressAtNodes ,StrainAtNodes, ...
        TpAtIntegrationPoints,VMAtIntegrationPoints, TpAtCentroid, TpAtNodes, VMAtNodes] = ...
            EvaluateFieldGradient(N_ELEM, N_NODE, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, Model2D,  UUR)

%% Get the Flux at the integration points and also the principal stresses
location = 1; % evaluate at the gauss points used for the solution
StressAtIntegrationPoints = zeros(N_ELEM*Ng*Ng,NDIM*NDIM);
StrainAtIntegrationPoints = zeros(N_ELEM*Ng*Ng,NDIM*NDIM);
TpAtIntegrationPoints = zeros(N_ELEM*Ng*Ng,3);
VMAtIntegrationPoints = zeros(N_ELEM*Ng*Ng,1);

for I_ELEM=1:N_ELEM
    [stress, strain] = GetElementFluxVector(I_ELEM, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, ... 
            Model2D,  location, UUR);
        
     e = ELEM_STIFF(I_ELEM,1);
     nu = ELEM_STIFF(I_ELEM,2);
     lambda = e*nu/((1+nu)*(1-2.*nu));
       
     for j=1:Ng*Ng
       pos =  (I_ELEM-1)*Ng*Ng + j;
       StressAtIntegrationPoints(pos,:)=stress(:,j);
       StrainAtIntegrationPoints(pos,:)=strain(:,j); 
       
       sigave = (stress(1,j)+stress(4,j))/2;
       radius = sqrt( ((stress(1,j)-stress(4,j))/2)^2 + stress(2,j)^2);
       
       TpAtIntegrationPoints(pos,1)=  sigave + radius;
       TpAtIntegrationPoints(pos,2)=  sigave - radius;
       if (Model2D == 1) 
            TpAtIntegrationPoints(pos,3)=0;
        elseif (Model2D == 2)        
            TpAtIntegrationPoints(pos,3)=lambda*(strain(1,j)+strain(4,j));
       end
        
       VMAtIntegrationPoints(pos) = sqrt( 0.5*(  (TpAtIntegrationPoints(pos,1)-TpAtIntegrationPoints(pos,2))^2 + ...
                                (TpAtIntegrationPoints(pos,1)-TpAtIntegrationPoints(pos,3))^2 + ...
                                (TpAtIntegrationPoints(pos,3)-TpAtIntegrationPoints(pos,2))^2   ) );
     end   
   
end

%% Get the Flux at the centroid for extrapolation to nodes
% Also evaluates the principal stresses and von mises stress at the
% centroid
location = 0; % evaluate at the center of the element
StrainAtCentroid = zeros(N_ELEM,NDIM*NDIM);
StressAtCentroid = zeros(N_ELEM,NDIM*NDIM);
TpAtCentroid = zeros(N_ELEM,3);

for I_ELEM=1:N_ELEM
    [stress, strain] = GetElementFluxVector(I_ELEM, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, ... 
            Model2D, location, UUR);
    StrainAtCentroid(I_ELEM,:)=strain'; 
    StressAtCentroid(I_ELEM,:)=stress'; 
    % principal stresses
    TpAtCentroid(I_ELEM,1)=(stress(1)+stress(4))/2 + sqrt( ((stress(1)-stress(4))/2)^2 + stress(2)^2);
    TpAtCentroid(I_ELEM,2)=(stress(1)+stress(4))/2 - sqrt( ((stress(1)-stress(4))/2)^2 + stress(2)^2);
    if (Model2D == 1) 
        TpAtCentroid(I_ELEM,3)=0;
    elseif (Model2D == 2)
        e = ELEM_STIFF(I_ELEM,1);
        nu = ELEM_STIFF(I_ELEM,2); 
        lambda = e*nu/((1+nu)*(1-2.*nu));
        TpAtCentroid(I_ELEM,3)=lambda*(strain(1)+strain(4));
    end
        
end

%% Extrapolate flux at nodes

scale = zeros(N_NODE,1); % vector containing the number of elements attached to each node
StressAtNodes = zeros(N_NODE,NDIM*NDIM); % matrix with the nodal values of the heat flux vector
StrainAtNodes = zeros(N_NODE,NDIM*NDIM);
TpAtNodes = zeros(N_NODE,3);
VMAtNodes = zeros(N_NODE,1);

for I_ELEM=1:N_ELEM
    for j=1:NNODE_ELE
        node = ELEM_NODE(j,I_ELEM);
        scale(node) = scale(node) + 1;
        StressAtNodes(node,:) = StressAtNodes(node,:) + StressAtCentroid(I_ELEM,:);
        StrainAtNodes(node,:)  = StrainAtNodes(node,:)  + StrainAtCentroid(I_ELEM,:);
        TpAtNodes(node,:) = TpAtNodes(node,:)+ TpAtCentroid(I_ELEM,:);
    end
end


for i=1:N_NODE
    for j=1:NDIM*NDIM
       StressAtNodes(i,j) = StressAtNodes(i,j)/scale(i);
       StrainAtNodes (i,j) = StrainAtNodes(i,j)/scale(i);       
    end
    TpAtNodes (i,:) = TpAtNodes(i,:)./scale(i);
    VMAtNodes(i) = sqrt( 0.5*(  (TpAtNodes(i,1)-TpAtNodes(i,2))^2 + ...
                                (TpAtNodes(i,1)-TpAtNodes(i,3))^2 + ...
                                (TpAtNodes(i,3)-TpAtNodes(i,2))^2   ) );
end

%%

File: 06-FEA-3D-Elasticity/C-Code/EvaluateFieldGradient.cpp
#include <MainFile.h>
using namespace std;


void EvaluateFieldGradient( const int nnode_ele, const int n_node, const int n_elem, const int dof_node,
		const int Ng, const int ShapeOrder, const int Model2D, const int ndim, const arma::mat& coords, 
		const arma::imat& elem_node, const arma::mat& elem_stiff, const arma::mat& Uur,
		arma::mat& StrainAtIntegrationPoints, arma::mat& StressAtIntegrationPoints,
		arma::mat& TpAtIntegrationPoints, arma::mat& VMAtIntegrationPoints,
		arma::mat& StrainAtCentroid, arma::mat& StressAtCentroid, arma::mat& TpAtCentroid,
		arma::mat& StrainAtNodes, arma::mat& StressAtNodes, arma::mat& TpAtNodes, arma::mat&  VMAtNodes)

{

	  StrainAtIntegrationPoints = arma::zeros<arma::mat>(n_elem*Ng*Ng,ndim*ndim);
	  StressAtIntegrationPoints = arma::zeros<arma::mat>(n_elem*Ng*Ng,ndim*ndim);
	  TpAtIntegrationPoints = arma::zeros<arma::mat>(n_elem*Ng*Ng,3);
	  VMAtIntegrationPoints = arma::zeros<arma::mat>(n_elem*Ng*Ng,1);
	  StrainAtCentroid = arma::zeros<arma::mat>(n_elem,ndim*ndim);
	  StressAtCentroid = arma::zeros<arma::mat>(n_elem,ndim*ndim);	
	  TpAtCentroid = arma::zeros<arma::mat>(n_elem,3);	
	  StrainAtNodes = arma::zeros<arma::mat>(n_node,ndim*ndim);
	  StressAtNodes = arma::zeros<arma::mat>(n_node,ndim*ndim);
	  TpAtNodes = arma::zeros<arma::mat>(n_node,3);
	  VMAtNodes = arma::zeros<arma::mat>(n_node,1);

	  // ------------------------------------------------------------
	  // ------- Get the Flux at the integration points -------------
	  // ------------------------------------------------------------
	  int location = 1; // evaluate at the gauss points used for the solution
	  arma::mat stress = arma::zeros<arma::mat>(ndim*ndim,Ng*Ng);
	  arma::mat strain = arma::zeros<arma::mat>(ndim*ndim,Ng*Ng);

	  for (int i=0;i<n_elem;i++) {
		  //cout<<"Element = "<<i<<"\n";
		  
		  double e = elem_stiff(i,0);
		  double nu = elem_stiff(i,1);
		  double lambda = e*nu/((1+nu)*(1-2.*nu));
		
		  int i_elem = i+1;
		  GetElementFluxVector(i_elem,nnode_ele, dof_node, Ng, ShapeOrder, Model2D, ndim, coords, elem_node,
					elem_stiff, location , Uur, stress, strain);
		  for (int j=0;j<Ng*Ng;j++){
			  int pos = (i)*Ng*Ng + j;
			  //cout<<"pos = "<<pos<<"\n";
			  StressAtIntegrationPoints.row(pos) = arma::trans(stress.col(j));
			  StrainAtIntegrationPoints.row(pos) = arma::trans(strain.col(j));
			  
			  double sigmaave = (stress(0,j) + stress(3,j))/2;
			  double radius = sqrt ( pow((stress(0,j) - stress(3,j))/2,2) + pow(stress(1,j),2)   );
			  
			  TpAtIntegrationPoints(pos,0) = sigmaave + radius;
			  TpAtIntegrationPoints(pos,1) = sigmaave - radius;
			  if (Model2D == 1) {
				  TpAtIntegrationPoints(pos,2) = 0;
			  }
			  else if(Model2D == 2) {
				  TpAtIntegrationPoints(pos,2) = lambda*(strain(0,j)+strain(3,j));
			  }
			  
			  VMAtIntegrationPoints(pos,0) = sqrt( 0.5* (
					  pow( TpAtIntegrationPoints(pos,0) - TpAtIntegrationPoints(pos,1),2) + 
					  pow( TpAtIntegrationPoints(pos,0) - TpAtIntegrationPoints(pos,2),2) + 
					  pow( TpAtIntegrationPoints(pos,1) - TpAtIntegrationPoints(pos,2),2) 
			  ));
			  
		  }
	  }
	  //cout<<StressAtIntegrationPoints<<"\n";
	  //cout<<StrainAtIntegrationPoints<<"\n";

	  // ------------------------------------------------------------
	  // ------- Get the Flux at the centroid -----------------------
	  // ------------------------------------------------------------
	  location = 0; // evaluate at the gauss points used for the solution
	  arma::mat stressC = arma::zeros<arma::mat>(ndim*ndim,1);
	  arma::mat strainC = arma::zeros<arma::mat>(ndim*ndim,1);

	  for (int i=0;i<n_elem;i++) {
		  //cout<<"Element = "<<i<<"\n";		  
		  double e = elem_stiff(i,0);
		  double nu = elem_stiff(i,1);
		  double lambda = e*nu/((1+nu)*(1-2.*nu));
		  int i_elem = i+1;
		  GetElementFluxVector(i_elem,nnode_ele, dof_node, Ng, ShapeOrder, Model2D, ndim, coords, elem_node,
					elem_stiff, location , Uur, stressC, strainC);
		  StressAtCentroid.row(i) = arma::trans(stressC);
		  StrainAtCentroid.row(i) = arma::trans(strainC);
		  
		  double sigmaave = (stressC(0) + stressC(3))/2;
		  double radius = sqrt ( pow((stressC(0) - stressC(3))/2,2) + pow(stressC(1),2)   );
		  
		  TpAtCentroid(i,0) = sigmaave + radius;
		  TpAtCentroid(i,1) = sigmaave - radius;
		  if (Model2D == 1) {
			  TpAtCentroid(i,2) = 0;
		  }
		  else if(Model2D == 2) {
			  TpAtCentroid(i,2) = lambda*(strainC(0)+strainC(3));
		  }
		  
		  
	  }


	  // ------------------------------------------------------------
	  // ------- Extrapolate flux to nodes    -----------------------
	  // ------------------------------------------------------------
	  arma::colvec scale = arma::zeros<arma::mat>(n_node);
	  int node;
	  for (int i=0;i<n_elem;i++) {
		  for (int j=0;j<nnode_ele;j++) {
			  node = elem_node(j,i)-1;
			  scale(node) = scale(node) + 1;
			  StressAtNodes.row(node) = StressAtNodes.row(node) + StressAtCentroid.row(i);
			  StrainAtNodes.row(node) = StrainAtNodes.row(node) + StrainAtCentroid.row(i);
			  TpAtNodes.row(node) = TpAtNodes.row(node) + TpAtCentroid.row(i);
		  }
	  }
	  for (int i=0;i<n_node;i++) {
	  		  for (int j=0;j<ndim*ndim;j++) {
	  			StressAtNodes(i,j) = StressAtNodes(i,j)/scale(i);
	  			StrainAtNodes(i,j) = StrainAtNodes(i,j)/scale(i);		  		
	  		  }
	  		  TpAtNodes.row(i) = TpAtNodes.row(i)/scale(i);
			  VMAtNodes(i) = sqrt( 0.5* (
					  pow( TpAtNodes(i,0) - TpAtNodes(i,1),2) + 
					  pow( TpAtNodes(i,0) - TpAtNodes(i,2),2) + 
					  pow( TpAtNodes(i,1) - TpAtNodes(i,2),2) 
			  ));
	  }



}
